File: src/config/column_config.py
"""
Dynamic column configuration for the model comparison table.
Defines column sets per model type plus defaults.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Callable, Dict, List, Optional


class ModelType(str, Enum):
    TEXT = "text"
    IMAGE = "image"
    VIDEO = "video"
    TTS = "tts"
    ASR = "asr"
    EMBEDDING = "embedding"
    UPSCALE = "upscale"
    INPAINT = "inpaint"


WIDTH_RESIZE = "resize_to_contents"
WIDTH_STRETCH = "stretch"


@dataclass(frozen=True)
class ColumnDefinition:
    """Defines a single column in the comparison table."""

    key: str
    header: str
    width_mode: str
    min_width: int = 60
    tooltip: Optional[str] = None
    sortable: bool = True
    formatter: Optional[Callable[..., object]] = None


TEXT_COLUMNS: List[ColumnDefinition] = [
    ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
    ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
    ColumnDefinition("context", "Context", WIDTH_RESIZE, 100),
    ColumnDefinition("quantization", "Quant", WIDTH_RESIZE, 70),
    ColumnDefinition("date_added", "Added", WIDTH_RESIZE, 90),
    ColumnDefinition("vision", "Vision", WIDTH_RESIZE, 70),
    ColumnDefinition("functions", "Functions", WIDTH_RESIZE, 85),
    ColumnDefinition("web_search", "Web Search", WIDTH_RESIZE, 90),
    ColumnDefinition("reasoning", "Reasoning", WIDTH_RESIZE, 85),
    ColumnDefinition("logprobs", "LogProbs", WIDTH_RESIZE, 80),
    ColumnDefinition("response_schema", "JSON", WIDTH_RESIZE, 55),
    ColumnDefinition("optimized_for_code", "Code Opt", WIDTH_RESIZE, 70),
    ColumnDefinition("audio_input", "Audio In", WIDTH_RESIZE, 75),
    ColumnDefinition("video_input", "Video In", WIDTH_RESIZE, 75),
    ColumnDefinition("input_price", "Input $/1M", WIDTH_RESIZE, 110),
    ColumnDefinition("output_price", "Output $/1M", WIDTH_STRETCH, 120),
    ColumnDefinition("cache_input", "Cache Read $/1M", WIDTH_RESIZE, 110),
    ColumnDefinition("cache_write", "Cache Write $/1M", WIDTH_RESIZE, 110),
    ColumnDefinition("privacy", "Privacy", WIDTH_RESIZE, 90),
]

IMAGE_COLUMNS: List[ColumnDefinition] = [
    ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
    ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
    ColumnDefinition("resolutions", "Resolutions", WIDTH_RESIZE, 150),
    ColumnDefinition("steps", "Steps", WIDTH_RESIZE, 80),
    ColumnDefinition("prompt_limit", "Prompt Limit", WIDTH_RESIZE, 110),
    ColumnDefinition("generation_price", "Price/Image", WIDTH_STRETCH, 110),
    ColumnDefinition("resolution_pricing", "Res Pricing", WIDTH_RESIZE, 110),
    ColumnDefinition("privacy", "Privacy", WIDTH_RESIZE, 90),
]

VIDEO_COLUMNS: List[ColumnDefinition] = [
    ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
    ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
    ColumnDefinition("video_type", "Video Type", WIDTH_RESIZE, 110),
    ColumnDefinition("durations", "Durations", WIDTH_RESIZE, 110),
    ColumnDefinition("resolutions", "Resolutions", WIDTH_RESIZE, 150),
    ColumnDefinition("aspect_ratios", "Aspect Ratios", WIDTH_RESIZE, 120),
    ColumnDefinition("audio", "Audio", WIDTH_RESIZE, 70),
    ColumnDefinition("audio_configurable", "Audio Config", WIDTH_RESIZE, 110),
    ColumnDefinition("base_price", "Base Price", WIDTH_RESIZE, 100),
    ColumnDefinition("audio_price", "Audio Price", WIDTH_RESIZE, 100),
    ColumnDefinition("privacy", "Privacy", WIDTH_RESIZE, 90),
]

TTS_COLUMNS: List[ColumnDefinition] = [
    ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
    ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
    ColumnDefinition("voices", "Voices", WIDTH_RESIZE, 110),
    ColumnDefinition("input_price", "Price/1M Chars", WIDTH_STRETCH, 130),
]

ASR_COLUMNS: List[ColumnDefinition] = [
    ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
    ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
    ColumnDefinition("input_price", "Price/Min", WIDTH_STRETCH, 110),
]

EMBEDDING_COLUMNS: List[ColumnDefinition] = [
    ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
    ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
    ColumnDefinition("dimensions", "Dimensions", WIDTH_RESIZE, 110),
    ColumnDefinition("input_price", "Input $/1M", WIDTH_STRETCH, 110),
]

UPSCALE_COLUMNS: List[ColumnDefinition] = [
    ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
    ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
    ColumnDefinition("upscale_factors", "Factors", WIDTH_RESIZE, 90),
    ColumnDefinition("upscale_price", "Price/Upscale", WIDTH_STRETCH, 120),
]

INPAINT_COLUMNS: List[ColumnDefinition] = [
    ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
    ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
    ColumnDefinition("resolutions", "Resolutions", WIDTH_RESIZE, 150),
    ColumnDefinition("steps", "Steps", WIDTH_RESIZE, 80),
    ColumnDefinition("inpaint_price", "Price/Inpaint", WIDTH_STRETCH, 120),
]

DEFAULT_COLUMNS: List[ColumnDefinition] = [
    ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
    ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
    ColumnDefinition("specs", "Specs", WIDTH_RESIZE, 150),
    ColumnDefinition("price", "Price", WIDTH_STRETCH, 110),
]

COLUMN_CONFIGS: Dict[ModelType, List[ColumnDefinition]] = {
    ModelType.TEXT: TEXT_COLUMNS,
    ModelType.IMAGE: IMAGE_COLUMNS,
    ModelType.VIDEO: VIDEO_COLUMNS,
    ModelType.TTS: TTS_COLUMNS,
    ModelType.ASR: ASR_COLUMNS,
    ModelType.EMBEDDING: EMBEDDING_COLUMNS,
    ModelType.UPSCALE: UPSCALE_COLUMNS,
    ModelType.INPAINT: INPAINT_COLUMNS,
}


def get_columns_for_type(model_type: Optional[str]) -> List[ColumnDefinition]:
    """Return column definitions for a given model type string."""
    if not model_type or model_type.lower() == "all":
        # For "all" or None, return common columns that work across all model types
        return [
            ColumnDefinition("model", "Model", WIDTH_RESIZE, 160),
            ColumnDefinition("type", "Type", WIDTH_RESIZE, 80),
            ColumnDefinition("specs", "Specs", WIDTH_RESIZE, 200),
            ColumnDefinition("price", "Price", WIDTH_STRETCH, 110),
        ]
    try:
        enum_type = ModelType(model_type.lower())
    except ValueError:
        return DEFAULT_COLUMNS
    return COLUMN_CONFIGS.get(enum_type, DEFAULT_COLUMNS)

File: web/components/models/columnConfig.ts
export interface ColumnDefinition {
  key: string
  header: string
  minWidth: number
  sortable: boolean
  tooltip?: string
}

export type ModelType = 'text' | 'image' | 'video' | 'tts' | 'asr' | 'embedding' | 'upscale' | 'inpaint' | 'all'

export const TEXT_COLUMNS: ColumnDefinition[] = [
  { key: 'model', header: 'Model', minWidth: 160, sortable: true },
  { key: 'type', header: 'Type', minWidth: 80, sortable: true },
  { key: 'context', header: 'Context', minWidth: 100, sortable: true, tooltip: 'Context window in tokens' },
  { key: 'quantization', header: 'Quant', minWidth: 70, sortable: true },
  { key: 'date_added', header: 'Added', minWidth: 90, sortable: true },
  { key: 'vision', header: 'Vision', minWidth: 70, sortable: true },
  { key: 'functions', header: 'Functions', minWidth: 85, sortable: true },
  { key: 'web_search', header: 'Web Search', minWidth: 90, sortable: true },
  { key: 'reasoning', header: 'Reasoning', minWidth: 85, sortable: true },
  { key: 'logprobs', header: 'LogProbs', minWidth: 80, sortable: true },
  { key: 'response_schema', header: 'JSON', minWidth: 55, sortable: true, tooltip: 'JSON response schema' },
  { key: 'optimized_for_code', header: 'Code Opt', minWidth: 70, sortable: true },
  { key: 'audio_input', header: 'Audio In', minWidth: 75, sortable: true },
  { key: 'video_input', header: 'Video In', minWidth: 75, sortable: true },
  { key: 'input_price', header: 'Input $/1M', minWidth: 110, sortable: true },
  { key: 'output_price', header: 'Output $/1M', minWidth: 120, sortable: true },
  { key: 'cache_input', header: 'Cache Read $/1M', minWidth: 110, sortable: true },
  { key: 'cache_write', header: 'Cache Write $/1M', minWidth: 110, sortable: true },
  { key: 'privacy', header: 'Privacy', minWidth: 90, sortable: true },
]

export const IMAGE_COLUMNS: ColumnDefinition[] = [
  { key: 'model', header: 'Model', minWidth: 160, sortable: true },
  { key: 'type', header: 'Type', minWidth: 80, sortable: true },
  { key: 'resolutions', header: 'Resolutions', minWidth: 150, sortable: true },
  { key: 'steps', header: 'Steps', minWidth: 80, sortable: true },
  { key: 'prompt_limit', header: 'Prompt Limit', minWidth: 110, sortable: true },
  { key: 'generation_price', header: 'Price/Image', minWidth: 110, sortable: true },
  { key: 'privacy', header: 'Privacy', minWidth: 90, sortable: true },
]

export const VIDEO_COLUMNS: ColumnDefinition[] = [
  { key: 'model', header: 'Model', minWidth: 160, sortable: true },
  { key: 'type', header: 'Type', minWidth: 80, sortable: true },
  { key: 'video_type', header: 'Video Type', minWidth: 110, sortable: true, tooltip: 'text-to-video or image-to-video' },
  { key: 'durations', header: 'Durations', minWidth: 110, sortable: true },
  { key: 'resolutions', header: 'Resolutions', minWidth: 150, sortable: true },
  { key: 'aspect_ratios', header: 'Aspect Ratios', minWidth: 120, sortable: true },
  { key: 'audio', header: 'Audio', minWidth: 70, sortable: true },
  { key: 'audio_configurable', header: 'Audio Config', minWidth: 110, sortable: true },
  { key: 'base_price', header: 'Base Price', minWidth: 100, sortable: true },
  { key: 'audio_price', header: 'Audio Price', minWidth: 100, sortable: true },
  { key: 'privacy', header: 'Privacy', minWidth: 90, sortable: true },
]

export const TTS_COLUMNS: ColumnDefinition[] = [
  { key: 'model', header: 'Model', minWidth: 160, sortable: true },
  { key: 'type', header: 'Type', minWidth: 80, sortable: true },
  { key: 'voices', header: 'Voices', minWidth: 110, sortable: true },
  { key: 'input_price', header: 'Price/1M Chars', minWidth: 130, sortable: true },
]

export const ASR_COLUMNS: ColumnDefinition[] = [
  { key: 'model', header: 'Model', minWidth: 160, sortable: true },
  { key: 'type', header: 'Type', minWidth: 80, sortable: true },
  { key: 'input_price', header: 'Price/Min', minWidth: 110, sortable: true },
]

export const EMBEDDING_COLUMNS: ColumnDefinition[] = [
  { key: 'model', header: 'Model', minWidth: 160, sortable: true },
  { key: 'type', header: 'Type', minWidth: 80, sortable: true },
  { key: 'dimensions', header: 'Dimensions', minWidth: 110, sortable: true },
  { key: 'input_price', header: 'Input $/1M', minWidth: 110, sortable: true },
]

export const UPSCALE_COLUMNS: ColumnDefinition[] = [
  { key: 'model', header: 'Model', minWidth: 160, sortable: true },
  { key: 'type', header: 'Type', minWidth: 80, sortable: true },
  { key: 'upscale_factors', header: 'Factors', minWidth: 90, sortable: true },
  { key: 'upscale_price', header: 'Price/Upscale', minWidth: 120, sortable: true },
]

export const INPAINT_COLUMNS: ColumnDefinition[] = [
  { key: 'model', header: 'Model', minWidth: 160, sortable: true },
  { key: 'type', header: 'Type', minWidth: 80, sortable: true },
  { key: 'resolutions', header: 'Resolutions', minWidth: 150, sortable: true },
  { key: 'steps', header: 'Steps', minWidth: 80, sortable: true },
  { key: 'inpaint_price', header: 'Price/Inpaint', minWidth: 120, sortable: true },
]

export const DEFAULT_COLUMNS: ColumnDefinition[] = [
  { key: 'model', header: 'Model', minWidth: 160, sortable: true },
  { key: 'type', header: 'Type', minWidth: 80, sortable: true },
  { key: 'specs', header: 'Specs', minWidth: 150, sortable: true },
  { key: 'price', header: 'Price', minWidth: 110, sortable: true },
]

const COLUMN_CONFIGS: Record<ModelType, ColumnDefinition[]> = {
  text: TEXT_COLUMNS,
  image: IMAGE_COLUMNS,
  video: VIDEO_COLUMNS,
  tts: TTS_COLUMNS,
  asr: ASR_COLUMNS,
  embedding: EMBEDDING_COLUMNS,
  upscale: UPSCALE_COLUMNS,
  inpaint: INPAINT_COLUMNS,
  all: DEFAULT_COLUMNS,
}

export function getColumnsForType(modelType: ModelType): ColumnDefinition[] {
  return COLUMN_CONFIGS[modelType] || DEFAULT_COLUMNS
}

export function getDefaultVisibleColumns(modelType: ModelType): string[] {
  const columns = getColumnsForType(modelType)
  return columns.map(c => c.key)
}

const STORAGE_KEY = 'vvv-model-column-preferences'

export function loadColumnPreferences(modelType: ModelType): Set<string> {
  if (typeof window === 'undefined') return new Set()
  try {
    const stored = localStorage.getItem(STORAGE_KEY)
    if (stored) {
      const data = JSON.parse(stored)
      const hidden = data[modelType] || []
      return new Set(hidden)
    }
  } catch {
    return new Set()
  }
  return new Set()
}

export function saveColumnPreferences(modelType: ModelType, hiddenColumns: Set<string>): void {
  if (typeof window === 'undefined') return
  try {
    const stored = localStorage.getItem(STORAGE_KEY)
    const data = stored ? JSON.parse(stored) : {}
    data[modelType] = Array.from(hiddenColumns)
    localStorage.setItem(STORAGE_KEY, JSON.stringify(data))
  } catch {
    return
  }
}

export type SortDirection = 'asc' | 'desc'

export interface SortConfig {
  key: string
  direction: SortDirection
}